pkg/iris/scan.go

```diff
@@ -67,64 +76,139 @@ func (s *scanner) Scan(config *rest.Config, workspaceId string, configurationNam
+	}
+
+	scannedObjectsByte, err := storage.Marshal(events)
+	if err != nil {
+		return s.LogAndShareError("Marshall scanned services", FAILED, err, kubernetesConfig.ID, workspaceId, accessToken)
+	}
+	err = s.irisApi.PostResults(scannedObjectsByte, accessToken)
+	if err != nil {
+		return s.LogAndShareError("Scan failed while posting results. RunId: [%s], with reason %v", FAILED, err, kubernetesConfig.ID, workspaceId, accessToken)
 	}
+
 	log.Infof("Scan Finished for RunId: [%s]", s.runId)
 	err = s.ShareStatus(kubernetesConfig.ID, workspaceId, accessToken, SUCCESSFUL, "Successfully Scanned")
 	if err != nil {
-		log.Errorf("Scan failed while posting status. RunId: [%s], with reason %v", s.runId, err)
+		log.Errorf(StatusErrorFormat, s.runId, err)
 		return err
 	}
-	return nil
+	return err
+}
+
+func (s scanner) ScanNamespace(k8sApi *kubernetes.API, mapper Mapper, namespaces []corev1.Namespace, cluster ClusterDTO, workspaceId string) ([]models.DiscoveryItem, error) {
+	// Metadata for the event
+	scope := fmt.Sprintf("workspace/%s", workspaceId)
+	source := fmt.Sprintf("kubernetes/%s#%s", cluster.name, s.runId)
+	var events []models.DiscoveryItem
+	for _, namespace := range namespaces {
+		// collect all deployments
+		deployments, err := k8sApi.Deployments(namespace.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		services, err := k8sApi.Services(namespace.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		mappedDeployments, err := mapper.MapDeployments(deployments, services)
+		if err != nil {
+			return nil, err
+		}
+
+		// create kubernetes event for namespace
+		discoveryEvent := s.CreateDiscoveryEvent(namespace, mappedDeployments, &cluster, source, scope)
+		events = append(events, discoveryEvent)
+	}
+
+	return events, nil
+}
+
+func (s scanner) LogAndShareError(message string, status string, err error, id string, workspaceId string, accessToken string) error {
+	log.Errorf(message, s.runId, err)
+	err = s.ShareStatus(id, workspaceId, accessToken, status, message)
+	if err != nil {
+		log.Errorf(StatusErrorFormat, s.runId, err)
+	}
+	return err
+}
+
+func (s *scanner) CreateDiscoveryEvent(namespace corev1.Namespace, deployments []models.Deployment, clusterDTO *ClusterDTO, source string, scope string) models.DiscoveryItem {
+	result := models.Cluster{
+		Namespace: models.Namespace{
+			Name: namespace.Name,
+		},
+		Deployments: deployments,
+		Name:        clusterDTO.name,
+		Os:          clusterDTO.osImage,
+		K8sVersion:  clusterDTO.k8sVersion,
+		NoOfNodes:   strconv.Itoa(clusterDTO.nodesCount),
+	}
+
+	// Metadata for the event
+
+	id := fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%s-%s", clusterDTO.name, namespace.Name))))
+	subject := fmt.Sprintf("namespace/%s", namespace.Name)
+	time := time2.Now().Format(time2.RFC3339)
+
+	// Build service/softwareArtifact event
+	discoveryEvent := New().
+		Id(id).
```

Comment:
Just out of curiosity, what is the advantage of using this event builder structure compared to simply creating  and returning the discoveryItem right here?

Reply:
Here it is a little bit over engineered but if some attributes become optional we can control this more easily with the builder pattern e.g.

Reply:
https://refactoring.guru/design-patterns/builder